### ISSUE TITLE
ZCL_ABAPGIT_REPO log refactoring

### ISSUE THREAD

**Issue by larshp** [larshp/abapGit#2821]
Suggest the following:
remove method CREATE_NEW_LOG
remove method GET_LOG
remove method RESET_LOG
remove instance variable MI_LOG

and instead have the log returned for the action that is performed, currently, this is just method DESERIALIZE.

**Comment by larshp** [comment]
this is probably difficult

**Comment by larshp** [comment]
@mfsap any thoughts? @ThomasPloski

**Comment by sbcgua** [comment]
Maybe (maybe) it makes sense to leave `get_log` as `get_last_log`. E.g. `export_zip` command is performed in the router (and called from 2 places) and may have some log which should be displayed somehow. Either in ALV `zcl_abapgit_log_viewer` or in html. The latter is rendered after so has no other way to get the log instance.

Not sure, just thoughts

**Comment by larshp** [comment]
I think the log should be returned to whatever performs the action, then its up to the caller to display/handle it

**Comment by sbcgua** [comment]
Yes, it would be better of course ... just not sure that it is always possible e.g. like in my example.

A question: `ZCL_ABAPGIT_REPO_SRV->ZIF_ABAPGIT_REPO_SRV~PURGE` why do we need `refresh` on failed deletion ? Or, more properly, why this refresh should happen so deeply ?

**Comment by larshp** [comment]
Could also leave it to the caller to refresh

I think its okay to refresh, to let the user see what the new status is

**Comment by sbcgua** [comment]
Looking at `ZCL_ABAPGIT_REPO_SRV->ZIF_ABAPGIT_REPO_SRV~PURGE` and `zcl_abapgit_objects=>delete` - the latter one collects exceptions for objects to the log. So in fact it builds it and is the only source of it.

What about carrying the log with exception (optionally) ? `zcx_abapgit_exception=>raise( iv_text = 'Error during uninstall. Check the log.' ii_log = li_log ).` ? I use this approach in one of my products. Not bad. Downsides: A) exception (which is a very very foundational primitive) starts knowing about more highlevel things like log (just an interface though) B) it needs that other parts of the code are aware of potential log in an exception - it took us looong time to make it proper in our case. But the result is quite convenient. @larshp ?

**Comment by larshp** [comment]
the log in abapGit also contains success and warning messages, I see exceptions as errors ie. a subset of the log functionality

**Comment by sbcgua** [comment]
Yes. But it can be transported differently. The top level code can receive it both as an exception and as a return param. Otherwise it has to catch exceptions and rethrow it.

Eg. the example above:
- `zcl_abapgit_objects=>delete` piles message in a log created in the caller and also can freely raise (line 607) - so mixed approach right already
- The caller - `ZCL_ABAPGIT_REPO_SRV->ZIF_ABAPGIT_REPO_SRV~PURGE` - creates the log, and also catches/rethrows exception to keep it in the common context (repo currently)
- The caller of purge - `ZCL_ABAPGIT_SERVICES_REPO=>PURGE` - does not catch anything and processes only the log received from `ZCL_ABAPGIT_REPO_SRV`

So currently it is very ununiform

There are also many methods which receive an instance of log (e.g. zcl_abapgit_serialize->files_local) and also return something. So in fact it "returns" the data and log but they are passed by different params. Which is also not ideal. Maybe it should return a structure of data and log. Or throw with log.

Just thoughts, no clear solution for me currently. In my product we had the same issue: log with errors and success and transferring of it from the bottom sometimes. Exceptions work. Though there are more wiring around and some key check points to popup that log in UI

**Comment by larshp** [comment]
avoiding rethrowing would be nice, perhaps change one of the places and open a PR? lets see how it looks

**Comment by sbcgua** [comment]
Done in #4590. 

One thing remaining is repo->get_log. Which is currently always uninitialized. Which should trigger "Last log" button in the toolbar. Which means that there should be some place for the last log ... I don't see a clear solution for the moment. Maybe in a separate PR :) Just one idea at the moment to keep it at page level actually via `zif_abapgit_gui_error_handler`

>> ZCL_ABAPGIT_REPO_SRV->ZIF_ABAPGIT_REPO_SRV~PURGE why do we need refresh on failed deletion ?

> I think its okay to refresh, to let the user see what the new status is

Currently removed the refresh at all, no clear place for it at the moment. Could be `CL_ABAPGIT_SERVICES_REPO=>PURGE` but also with rethrowing (better than in repo_srv but ...)

**Comment by larshp** [comment]
the UI/page should keep the log, if its to be shown to the user multiple times after navigation

the log does not necessarily contain an error, so I'd not expect logs to be kept in `zif_abapgit_gui_error_handler`

**Comment by larshp** [comment]
still relevant
